### PR TITLE
[nemo-qml-plugin-email] Don't set URL in attachment list model if the file does not exist. Fixes JB#56042

### PR DESCRIPTION
**Pull request by dcaliste**
… file does not exist.

@pvuorela, I'm trying to solve https://forum.sailfishos.org/t/email-attachment-opening-attachment-display-inode-error/4805/6

The reproducer is the following:
- use POP
- get an email with at least two attached files
- tap on the first one : it is opened
- tap on the second, a banner will complain about the impossibility to open inode mime type

and bug path is:
- in attachmentlistmodel, after the first file has been copied from QMF internal storage to the file structure in Download, the `QFileSystemWatcher` callback line 298 in triggered. But for other attachment, the files have not been copied yet. So `attachmentPath()` is return an empty string,
- then `onAttachmentPathChanged()` is creating an URL with just the directory.

I think the `onAttachmentPathChanged()` should not mangle the path. It is called from the email agent that is getting the path from QMF only and thus guaranteed to be valid. The fact that we are using this callback by hand in the file watcher with a wrong path is the problem.

The PR here is fixing the problem, but I would consider two better possibilities for medium term :
- get the filename for a part directly from QMF instead of mimicking its behaviour with `.remove('/')` statements,
- remove the directory watcher completely, since the email agent will signal anyway when a file has been downloaded and copied to a valid URL.

### CODE REVIEW

File: src/attachmentlistmodel.cpp [hunk 1]
```diff
@@ -148,15 +148,12 @@ void AttachmentListModel::onAttachmentDownloadProgressChanged(const QString &att
...
         Attachment *attachment = m_attachmentsList.at(i);
         if (attachment->location == attachmentLocation) {
-            // Make url more secure and compatible with QMF client
-            QString downloads = downloadFolder(m_message, attachmentLocation);
-            QString fileName = path;
-            QString secureUrl = QUrl::fromLocalFile(downloads + "/" + fileName.remove(downloads).remove('/')).toString();
-            if (attachment->url != secureUrl) {
-                attachment->url = secureUrl;
+            if (attachment->url != path) {
+                attachment->url = path;
```
Comment (pvuorela) [inline src/attachmentlistmodel.cpp:156]: So effectively reverting part of commit bddff0f196d4 here. 

The commit message on that was somewhat short but the problem was that the email could claim some "../../foo" type of filename in the content and the the code in this class was reading the file elsewhere. 

But indeed looks like the path here is coming from either EmailAgent after saving something or from the QFileSystemWatcher here (as in ::resetMode()) which should(?) filter the result with attachmentPath().  cc @mkenttala and @llewelld

**Comment by llewelld** [comment]
Pekka is correct: this change was intended to catch nefarious attachment names coming from QMF (`sourcePart.displayName()` isn't safe as it comes from the email). The key aim is to ensure that whatever comes from QMF, the file is never saved outside the Downloads folder. Making this reversion at this point in time would be problematic unfortunately.

This should also be fixed in QMF (i.e. to prevent it passing nefarious paths), but for various reasons (defence in depth, QMF's plugin structure), I'd argue it still makes sense to sanitise the path here as well.

Could we use some mangling like this instead, to catch the empty case?
```
QString mangled = !path.isEmpty() ? downloads + QDir::separator() + QUrl(path).fileName() : QString();
```

**Comment by llewelld** [comment]
Just to add, I just realised that mangling has to match up with the mangling in QMF, which won't be the case for my suggestion. But hopefully the idea is clear.

**Comment by dcaliste** [comment]
I think that in QMF, all '/' characters are removed from the path, see qmailmessage.cpp#6349. Thus, displayName, even if malicious could not lead to a path outside the intented directory for saving. That's from reading the code, I didn't test to create such a situation, so I may be wrong.

I'm sorry, but I still don't see how the part I reverted is correcting a bug. The only places this function is called from are:
* the callbacks from the agent signaling, thus coming with a path that has been sanitized by QMF (assuming that the sanitizing there is strong enough),
* a local callback triggered by directory changes where the file is supposed to be saved, and in this callback, the path is forged by `::attachmentUrl()` that is also removing '/' characters like in QMF.

So as far as I understand it, all code path arriving to the reverted code cannot contain something like "../../".

Maybe I'm missing something. Anyway, I can amend the PR here not to touch this part, as I mentioned on IRC with @pvuorela, the only relevant part if the added `if (!path.isEmpty())` in the directory callback.

**Comment by pvuorela** [comment]
> I'm sorry, but I still don't see how the part I reverted is correcting a bug

Yea the question being whether ::attachmentUrl() part is enough.

**Comment by llewelld** [comment]
Perhaps I'm misreading the code, but when the attachment is saved, `EmailAgent::saveAttachmentToDownloads()` will trigger a call to `AttachmentListModel::onAttachmentPathChanged()` with `path` set to `attachmentDownloadFolder + "/" + attachmentPart.displayName()`.

> Thus, displayName, even if malicious could not lead to a path outside the intented directory for saving.

That's correct, but it can lead to a path outside the intended directory for reading, which is also a security problem.

> thus coming with a path that has been sanitized by QMF 

True, but this is relying on an undocumented assumption about how the QMF library behaves, so there's also an element of defence in depth here too.

**Comment by dcaliste** [comment]
> trigger a call to AttachmentListModel::onAttachmentPathChanged() with path set to attachmentDownloadFolder + "/" + attachmentPart.displayName().

Ah, I see, it's the code path where file exist in emailagent that is problematic, on line 1452. Well I may argue that it's not really an issue since in that case, nothing is written since the file is assumed to exists already. So there is no security issue of overwriting a system file or something. But the display attachment will be rubbish. Thank you for pointing my error.

> True, but this is relying on an undocumented assumption about how the QMF library behaves,

This is a very valid concern in my opinion. The solution would be to expose the naming scheme from QMF. So let's keep this for another PR.

I'm amending the PR here just to keep the added `if (!path.isEmpty())`.

**Comment by llewelld** [comment]
> But the display attachment will be rubbish.

Unfortunately it could be worse than rubbish: it might be a privileged file that the user could then be tricked into forwarding. It's not the most likely scenario, but still one I think better not to expose.

It's a bit frustrating, I agree, and I also agree with your original point that it would be better to get the file name directly from QMF, rather than trying to reconstruct it.

**Review by llewelld — APPROVED** [review]
LGTM and works nicely; thank you for fixing this Damien.

**Review by pvuorela — APPROVED** [review]

**Pull request merged**